File: src/systems/WorldObjectSystem.js
/**
 * WorldObjectSystem - Handles interactions with world objects
 * Manages doors, furniture, and other interactive world elements
 */
export class WorldObjectSystem {
    constructor(game) {
        this.game = game;
    }
    
    /**
     * Perform action on a world object
     */
    performAction(worldObject, action, player, tool = null) {
        switch (action) {
            case 'open':
                return this.openDoor(worldObject, player);
            case 'close':
                return this.closeDoor(worldObject, player);
            case 'peek':
                return this.peekThroughDoor(worldObject, player);
            case 'knock':
                return this.knockOnDoor(worldObject, player);
            case 'smash':
                return this.smashObject(worldObject, player, tool);
            case 'search':
                return this.searchFurniture(worldObject, player);
            case 'use':
                return this.useObject(worldObject, player, tool);
            case 'disassemble':
                return this.disassembleObject(worldObject, player, tool);
            case 'lockpick':
                return this.lockpickDoor(worldObject, player, tool);
            case 'barricade':
                return this.barricadeDoor(worldObject, player);
            case 'remove_barricade':
                return this.removeBarricade(worldObject, player);
            default:
                return { success: false, message: 'Unknown action.' };
        }
    }
    
    /**
     * Open a door
     */
    openDoor(door, player) {
        const result = door.open();
        
        if (result.success) {
            this.game.world.updateTileAt(door.x, door.y, door.z, door.getTile());
            this.game.advanceTurn(result.timeSpent || 1);
        }
        
        return result;
    }
    
    /**
     * Close a door
     */
    closeDoor(door, player) {
        const result = door.close();
        
        if (result.success) {
            this.game.world.updateTileAt(door.x, door.y, door.z, door.getTile());
            this.game.advanceTurn(result.timeSpent || 1);
        }
        
        return result;
    }
    
    /**
     * Peek through a door
     */
    peekThroughDoor(door, player) {
        const result = door.peek();
        
        if (result.success) {
            this.game.advanceTurn(result.timeSpent || 1);
            
            // TODO: Show limited vision through door
            // For now, just reveal tiles immediately adjacent on other side
            const dx = door.x - player.x;
            const dy = door.y - player.y;
            
            // Peek one tile beyond the door
            const peekX = door.x + (dx !== 0 ? Math.sign(dx) : 0);
            const peekY = door.y + (dy !== 0 ? Math.sign(dy) : 0);
            
            const tile = this.game.world.getTileAt(peekX, peekY, door.z);
            if (tile) {
                result.message += ` You see: ${tile.name || 'floor'}.`;
            }
        }
        
        return result;
    }
    
    /**
     * Knock on a door
     */
    knockOnDoor(door, player) {
        const result = door.knock();
        
        if (result.success) {
            this.game.advanceTurn(result.timeSpent || 1);
            
            // Make noise to alert NPCs
            if (result.makeNoise) {
                this.makeNoise(door.x, door.y, door.z, result.makeNoise);
            }
        }
        
        return result;
    }
    
    /**
     * Smash an object (door, furniture, etc.)
     * Auto-completes: loops hits until destroyed or weapon breaks
     */
    smashObject(object, player, weapon) {
        // Get weapon from player's hands if not specified
        if (!weapon) {
            weapon = player.equipment.rightHand || player.equipment.leftHand;
        }
        
        // Allow unarmed kicking for weak materials
        let isKicking = false;
        if (!weapon) {
            // Check if material is weak enough to kick
            const canKick = object.material === 'wood' || 
                           object.material === 'glass';
            
            if (!canKick) {
                return { success: false, message: 'You need a weapon or tool to smash this. Metal doors are too strong to kick.' };
            }
            
            // Create fake "kick" weapon
            weapon = { name: 'kick', damage: 8, tags: ['blunt'] };
            isKicking = true;
        }
        
        // Pre-calculate per-hit values
        const baseDamage = weapon.damage || 5;
        const materialMod = this.getMaterialModifier(object.material, weapon);
        const damagePerHit = baseDamage * materialMod;
        const timePerHit = Math.max(2, Math.ceil(5 - (weapon.damage || 0) / 10));
        
        let totalHits = 0;
        let totalTime = 0;
        let destroyed = false;
        let weaponBroke = false;
        let lockBroke = false;
        
        // Loop: keep hitting until destroyed or weapon breaks
        while (!destroyed) {
            // Apply damage to object
            const damageResult = object.takeDamage(damagePerHit);
            totalHits++;
            totalTime += timePerHit;
            
            // Apply durability damage to weapon (not for kicking)
            if (!isKicking && weapon.durability !== undefined) {
                weapon.durability = Math.max(0, weapon.durability - 2);
                if (weapon.durability <= 0) {
                    weaponBroke = true;
                }
            }
            
            // Make noise each hit
            this.makeNoise(object.x, object.y, object.z, {
                volume: 30,
                range: 20,
                type: 'smash',
                description: 'loud smashing'
            });
            
            // Check if lock broke mid-smash
            if (!lockBroke && object.state && object.state.locked && object.hp < object.maxHP * 0.3) {
                object.state.locked = false;
                lockBroke = true;
            }
            
            if (damageResult.destroyed) {
                destroyed = true;
            } else if (weaponBroke) {
                // Weapon broke before object was destroyed — stop
                break;
            }
        }
        
        // Build summary message
        const actionVerb = isKicking ? 'kick' : 'smash';
        let message = '';
        
        if (destroyed) {
            message = isKicking
                ? `You kick apart the ${object.name} in ${totalHits} hit${totalHits > 1 ? 's' : ''} (${totalTime} turns).`
                : `You smash the ${object.name} to pieces with your ${weapon.name} in ${totalHits} hit${totalHits > 1 ? 's' : ''} (${totalTime} turns).`;
            
            object.updateVisuals();
            this.game.world.updateTileAt(object.x, object.y, object.z, object.getTile());
            
            // Drop materials (reduced yield from smashing)
            const materials = this.rollDropTable(object.dropTable, 0.5);
            this.dropMaterials(materials, object.x, object.y, object.z);
            
            if (materials.length > 0) {
                message += ` Some materials scatter on the ground.`;
            }
            
            // Spill stored contents on ground if furniture was a container
            if (object.isContainer && object.pockets) {
                let spillCount = 0;
                for (const pocket of object.pockets) {
                    if (pocket.contents) {
                        for (const item of pocket.contents) {
                            item.x = object.x;
                            item.y = object.y;
                            item.z = object.z;
                            this.game.world.addItem(item);
                            spillCount++;
                        }
                        pocket.contents = [];
                    }
                }
                if (spillCount > 0) {
                    message += ` ${spillCount} item(s) spill out onto the ground.`;
                }
            }
        } else {
            // Weapon broke before finishing
            message = isKicking
                ? `You kick the ${object.name} ${totalHits} time${totalHits > 1 ? 's' : ''} (${totalTime} turns).`
                : `You ${actionVerb} the ${object.name} ${totalHits} time${totalHits > 1 ? 's' : ''} but your ${weapon.name} broke!`;
            message += ` (${Math.floor(object.hp)}/${object.maxHP} HP remaining)`;
            object.updateVisuals();
            this.game.world.updateTileAt(object.x, object.y, object.z, object.getTile());
        }
        
        if (lockBroke) {
            this.game.ui.log(`The lock on the ${object.name} breaks!`, 'info');
        }
        
        if (weaponBroke) {
            this.game.ui.log(`Your ${weapon.name} broke!`, 'warning');
        }
        
        this.game.advanceTurn(totalTime);
        
        return {
            success: true,
            message,
            destroyed,
            hits: totalHits,
            timeSpent: totalTime
        };
    }
    
    /**
     * Disassemble an object carefully
     */
    disassembleObject(object, player, tool) {
        // Check if object can be disassembled
        if (!object.dropTable || !object.dropTable.disassembleTool) {
            return { success: false, message: 'This cannot be disassembled.' };
        }
        
        // Check for required tool
        const requiredTool = object.dropTable.disassembleTool;
        if (!tool) {
            tool = this.findTool(player, requiredTool);
        }
        
        if (!tool) {
            return { 
                success: false, 
                message: `You need a ${requiredTool} to disassemble this.` 
            };
        }
        
        // Doors must be open to disassemble
        if (object.type === 'door' && !object.state.open) {
            return { success: false, message: `You must open the ${object.name} first.` };
        }
        
        // Time required based on object HP
        const timeSpent = Math.max(5, Math.ceil(object.maxHP / 10));
        
        // Get full materials (100% yield for careful disassembly)
        const materials = this.rollDropTable(object.dropTable, 1.0);
        
        // Remove object
        object.hp = 0;
        object.updateVisuals();
        this.game.world.updateTileAt(object.x, object.y, object.z, object.getTile());
        
        // Drop materials
        this.dropMaterials(materials, object.x, object.y, object.z);
        
        // Spill stored contents on ground if furniture was a container
        let spillMsg = '';
        if (object.isContainer && object.pockets) {
            let spillCount = 0;
            for (const pocket of object.pockets) {
                if (pocket.contents) {
                    for (const item of pocket.contents) {
                        item.x = object.x;
                        item.y = object.y;
                        item.z = object.z;
                        this.game.world.addItem(item);
                        spillCount++;
                    }
                    pocket.contents = [];
                }
            }
            if (spillCount > 0) {
                spillMsg = ` ${spillCount} stored item(s) dropped on ground.`;
            }
        }
        
        // Small durability damage to tool
        if (tool.durability !== undefined) {
            tool.durability = Math.max(0, tool.durability - 1);
        }
        
        this.game.advanceTurn(timeSpent);
        
        return {
            success: true,
            message: `You carefully disassemble the ${object.name}. (${materials.length} materials recovered)${spillMsg}`,
            materials
        };
    }
    
    /**
     * Lockpick a door (future feature)
     */
    lockpickDoor(door, player, tool) {
        // TODO: Implement lockpicking system
        return { 
            success: false, 
            message: 'Lockpicking not yet implemented. Try smashing the door instead.' 
        };
    }
    
    /**
     * Barricade a door (future feature)
     */
    barricadeDoor(door, player) {
        // TODO: Implement barricading system
        return { 
            success: false, 
            message: 'Barricading not yet implemented.' 
        };
    }
    
    /**
     * Remove barricade from door (future feature)
     */
    removeBarricade(door, player) {
        // TODO: Implement barricade removal
        return { 
            success: false, 
            message: 'Barricade removal not yet implemented.' 
        };
    }
    
    /**
     * Search furniture - marks as searched and signals UI to show contents
     */
    searchFurniture(furniture, player) {
        if (!furniture.isContainer) {
            return { success: false, message: 'Nothing to search here.' };
        }
        
        furniture.state.searched = true;
        this.game.advanceTurn(1);
        
        const itemCount = furniture.getStoredItemCount ? furniture.getStoredItemCount() : 0;
        
        return {
            success: true,
            message: `You search the ${furniture.name}. ${itemCount > 0 ? `Found ${itemCount} item(s).` : 'It\'s empty.'}`,
            showContents: true,
            worldObject: furniture
        };
    }
    
    /**
     * Make noise to alert NPCs
     */
    makeNoise(x, y, z, noiseData) {
        // Check if noise system exists
        if (!this.game.world.noiseSystem) {
            console.warn('Noise system not found - creating basic implementation');
            // Create basic noise alert for NPCs
            const entities = this.game.world.getAllEntities();
            for (const entity of entities) {
                if (entity.z !== z) continue;
                
                const distance = Math.sqrt(
                    Math.pow(entity.x - x, 2) + 
                    Math.pow(entity.y - y, 2)
                );
                
                if (distance <= noiseData.range) {
                    // Alert NPC
                    if (entity.onNoiseHeard) {
                        entity.onNoiseHeard(x, y, noiseData);
                    }
                }
            }
            
            this.game.ui.log(`You hear ${noiseData.description || 'a noise'}.`, 'warning');
        } else {
            // Use existing noise system
            this.game.world.noiseSystem.makeNoise(x, y, z, noiseData);
        }
    }
    
    /**
     * Get material modifier for damage calculation
     */
    getMaterialModifier(material, weapon) {
        const weaponType = weapon.tags || [];
        
        // Blunt weapons good against wood, bad against metal
        if (weaponType.includes('blunt')) {
            if (material === 'wood') return 1.5;
            if (material === 'metal') return 0.5;
            if (material === 'glass') return 2.0;
        }
        
        // Sharp weapons decent against everything
        if (weaponType.includes('sharp')) {
            if (material === 'wood') return 1.2;
            if (material === 'metal') return 0.7;
            if (material === 'glass') return 1.5;
        }
        
        // Default
        return 1.0;
    }
    
    /**
     * Roll drop table for materials
     */
    rollDropTable(dropTable, yieldModifier = 1.0) {
        if (!dropTable || !dropTable.materials) return [];
        
        const materials = [];
        
        for (const materialDef of dropTable.materials) {
            const quantity = Array.isArray(materialDef.quantity) 
                ? Math.floor(Math.random() * (materialDef.quantity[1] - materialDef.quantity[0] + 1)) + materialDef.quantity[0]
                : materialDef.quantity;
            
            const quality = Array.isArray(materialDef.quality)
                ? Math.floor(Math.random() * (materialDef.quality[1] - materialDef.quality[0] + 1)) + materialDef.quality[0]
                : materialDef.quality || 100;
            
            const actualQuantity = Math.max(1, Math.floor(quantity * yieldModifier));
            
            if (actualQuantity > 0) {
                materials.push({
                    name: materialDef.name,
                    quantity: actualQuantity,
                    quality: Math.floor(quality * yieldModifier)
                });
            }
        }
        
        return materials;
    }
    
    /**
     * Drop materials on ground
     */
    dropMaterials(materials, x, y, z) {
        const player = this.game.player;
        
        // Map drop-table display names → component IDs in ContentManager
        const DROP_NAME_TO_COMPONENT = {
            'Glass Shards': 'glass_shard',
            'Glass Shard':  'glass_shard',
            'Metal Scraps': 'scrap_metal_shard',
            'Metal Shard':  'scrap_metal_shard',
            'Wood Plank':   'wood_plank',
            'Wood Piece':   'wood_piece',
            'Nails':        'nail',
            'Nail':         'nail',
            'Screws':       'screw',
            'Screw':        'screw',
            'Fabric Panel': 'fabric_panel',
            'Pipe':         'metal_tube',
            'Bone Shard':   'bone_shard',
            'Rubber Piece': 'rubber_piece',
            'Duct Tape':    'duct_tape',
            'Cloth Wrap':   'cloth_wrap',
            'Stone':        'stone'
        };
        
        for (const material of materials) {
            // Try to create a proper component via ContentManager
            const componentId = DROP_NAME_TO_COMPONENT[material.name];
            let materialItem = null;
            
            if (componentId && this.game.content) {
                materialItem = this.game.content.createComponent(componentId);
            }
            
            if (materialItem) {
                // Override quantity/quality from the drop table roll
                materialItem.quantity = material.quantity;
                materialItem.quality = material.quality || 100;
                materialItem.maxQuality = 100;
            } else {
                // Fallback: create bare item for unknown materials (e.g. Steel Plate)
                materialItem = {
                    id: `${material.name.toLowerCase().replace(/\s+/g, '_')}_${Date.now()}_${Math.random()}`,
                    name: material.name,
                    type: 'component',
                    glyph: '*',
                    color: '#ffaa00',
                    weight: material.quantity * 10,
                    volume: material.quantity * 5,
                    quality: material.quality || 100,
                    maxQuality: 100,
                    quantity: material.quantity,
                    stackable: true,
                    tags: ['material', 'salvage'],
                    isComponent: true
                };
            }
            
            materialItem.x = x;
            materialItem.y = y;
            materialItem.z = z;
            
            // Try to add to player inventory first
            const addResult = player.addToInventory(materialItem);
            if (!addResult.success) {
                // Inventory full, drop on ground
                this.game.world.addItem(materialItem);
                this.game.ui.log(`${material.name} x${material.quantity} dropped on ground (no storage space).`, 'warning');
            } else {
                this.game.ui.log(`Recovered ${material.name} x${material.quantity}.`, 'info');
            }
        }
    }
    
    /**
     * Find tool in player inventory/equipment
     */
    findTool(player, toolType) {
        // Check equipment
        for (const slot in player.equipment) {
            const item = player.equipment[slot];
            if (!item) continue;
            
            if (this.isToolType(item, toolType)) {
                return item;
            }
        }
        
        // Check inventory
        for (const item of player.inventory) {
            if (this.isToolType(item, toolType)) {
                return item;
            }
        }
        
        return null;
    }
    
    /**
     * Check if item matches tool type
     */
    isToolType(item, toolType) {
        if (!item.tags) return false;
        
        const toolMap = {
            'screwdriver': ['screwdriver', 'multitool'],
            'crowbar': ['crowbar', 'prybar'],
            'hacksaw': ['hacksaw', 'saw'],
            'lockpick': ['lockpick', 'lockpick_set']
        };
        
        const validTags = toolMap[toolType] || [toolType];
        
        return validTags.some(tag => item.tags.includes(tag));
    }
}

File: Nightrun_v01/Systems/WorldObjectSystem.cs
using Nightrun.Content;
using Nightrun.Entities;
using Nightrun.World;

namespace Nightrun.Systems;

/// <summary>
/// Outcome of a world-object action. <see cref="TimeCost"/> is how many turns
/// the caller should tick; <see cref="Destroyed"/> is true when the object
/// was reduced to 0 HP (smash/disassemble).
/// </summary>
public readonly struct WorldObjectResult
{
    public readonly bool   Success;
    public readonly string Message;
    public readonly int    TimeCost;
    public readonly bool   Destroyed;
    public readonly int    NoiseVolume;
    public readonly int    NoiseRange;

    public WorldObjectResult(bool success, string message, int timeCost = 0,
                             bool destroyed = false,
                             int noiseVolume = 0, int noiseRange = 0)
    {
        Success = success; Message = message; TimeCost = timeCost;
        Destroyed = destroyed; NoiseVolume = noiseVolume; NoiseRange = noiseRange;
    }

    public static WorldObjectResult Fail(string msg) => new(false, msg);
}

/// <summary>
/// Port of <c>src/systems/WorldObjectSystem.js</c> — the "core" subset:
/// peek, knock, smash (with material drops), and careful disassembly of
/// adjacent world objects. Lockpick / barricade are stubbed for later.
/// Noise generation returns a (volume,range) pair in the result so the
/// future <see cref="SoundSystem"/> can consume it without this class having
/// to depend on it yet.
/// </summary>
public sealed class WorldObjectSystem
{
    private readonly IWorldMap _world;
    private readonly Random _rng;
    private readonly Action<string> _log;

    public WorldObjectSystem(IWorldMap world, Random rng, Action<string> log)
    {
        _world = world; _rng = rng; _log = log;
    }

    // ------------------------------------------------------------------
    // Peek / Knock — both are closed-door-only and cheap.

    public WorldObjectResult Peek(Door door, Player player)
    {
        if (door.IsDestroyed) return WorldObjectResult.Fail($"The {door.Name.ToLower()} is gone.");
        if (door.Open)        return WorldObjectResult.Fail($"The {door.Name.ToLower()} is already open.");

        // Reveal one tile on the far side.
        int dx = Math.Sign(door.X - player.X);
        int dy = Math.Sign(door.Y - player.Y);
        int px = door.X + dx;
        int py = door.Y + dy;
        var tile = _world.GetTile(px, py, door.Z);
        var def  = TileCatalog.Get(tile.Type);

        string what = def.Name;
        var obj = _world.GetObjectAt(px, py, door.Z);
        if (obj != null) what = obj.Name.ToLower();
        else
        {
            var items = _world.GetItemsAt(px, py, door.Z);
            if (items != null && items.Count > 0) what = items[items.Count - 1].Name.ToLower();
            var npc = _world.GetNpcAt(px, py, door.Z);
            if (npc != null && !npc.IsDead) what = npc.Name.ToLower();
        }

        return new WorldObjectResult(
            success: true,
            message: $"You peek through the {door.Name.ToLower()}. You see: {what}.",
            timeCost: 1);
    }

    public WorldObjectResult Knock(Door door, Player player)
    {
        if (door.IsDestroyed) return WorldObjectResult.Fail($"The {door.Name.ToLower()} is gone.");
        if (door.Open)        return WorldObjectResult.Fail($"The {door.Name.ToLower()} is already open.");

        return new WorldObjectResult(
            success: true,
            message: $"You knock on the {door.Name.ToLower()}.",
            timeCost: 1,
            noiseVolume: 15,
            noiseRange: 10);
    }

    // ------------------------------------------------------------------
    // Smash — loops hits until destroyed or the weapon breaks. Damage per
    // hit is the average of the weapon's dice (or a small fallback for
    // unarmed kicking against weak materials). Port of
    // WorldObjectSystem.smashObject.

    public WorldObjectResult Smash(WorldObject obj, Player player)
    {
        if (obj.IsDestroyed)
            return WorldObjectResult.Fail($"The {obj.Name.ToLower()} is already destroyed.");

        var weapon = player.Equipment.RHand ?? player.Equipment.LHand;
        bool kicking = false;
        string weaponName;
        int baseDamage;
        List<string> weaponTags;

        if (weapon == null)
        {
            // Only weak materials can be kicked.
            if (obj.Material != "wood" && obj.Material != "glass")
                return WorldObjectResult.Fail(
                    "You need a weapon or tool to smash this. It's too strong to kick.");
            kicking     = true;
            weaponName  = "kick";
            baseDamage  = 8;
            weaponTags  = new List<string> { "blunt" };
        }
        else
        {
            weaponName = weapon.Name;
            baseDamage = AverageDice(weapon.DamageDice) + (weapon.AttackType == "blunt" ? 1 : 0);
            if (baseDamage < 1) baseDamage = 5;
            weaponTags = weapon.Tags;
        }

        double matMod      = GetMaterialModifier(obj.Material, weaponTags);
        int    damagePerHit = Math.Max(1, (int)(baseDamage * matMod));
        int    timePerHit   = Math.Max(2, (int)Math.Ceiling(5 - baseDamage / 10.0));

        int hits = 0, totalTime = 0;
        bool destroyed = false, weaponBroke = false, lockBroke = false;

        while (!destroyed)
        {
            obj.TakeDamage(damagePerHit);
            hits++;
            totalTime += timePerHit;

            if (!kicking && weapon != null)
            {
                weapon.Durability = Math.Max(0, weapon.Durability - 2);
                if (weapon.Durability <= 0) weaponBroke = true;
            }

            // Lock breaks at 30% HP — auto-unlocks the door.
            if (!lockBroke && obj is Door d && d.Locked && obj.HP < obj.MaxHP * 0.3)
            {
                d.Locked = false;
                lockBroke = true;
            }

            if (obj.IsDestroyed) { destroyed = true; break; }
            if (weaponBroke) break;
        }

        string verb = kicking ? "kick" : "smash";
        string message;
        if (destroyed)
        {
            message = kicking
                ? $"You kick apart the {obj.Name.ToLower()} in {hits} hit{(hits > 1 ? "s" : "")} ({totalTime} turns)."
                : $"You smash the {obj.Name.ToLower()} to pieces with your {weaponName} in {hits} hit{(hits > 1 ? "s" : "")} ({totalTime} turns).";

            ApplyDestruction(obj);
            var mats = RollDropTable(DropTableCatalog.For(obj), yieldMod: 0.5);
            int dropped = DropMaterials(mats, obj.X, obj.Y, obj.Z, player);
            if (dropped > 0) message += " Some materials scatter on the ground.";

            int spilled = SpillContents(obj);
            if (spilled > 0) message += $" {spilled} item(s) spill out onto the ground.";
        }
        else
        {
            message = kicking
                ? $"You kick the {obj.Name.ToLower()} {hits} time{(hits > 1 ? "s" : "")} ({totalTime} turns)."
                : $"You {verb} the {obj.Name.ToLower()} {hits} time{(hits > 1 ? "s" : "")} but your {weaponName} broke!";
            message += $" ({obj.HP}/{obj.MaxHP} HP remaining)";
        }

        if (lockBroke) _log($"The lock on the {obj.Name.ToLower()} breaks!");
        if (weaponBroke) _log($"Your {weaponName} broke!");

        return new WorldObjectResult(
            success: true,
            message: message,
            timeCost: totalTime,
            destroyed: destroyed,
            noiseVolume: 30,
            noiseRange: 20);
    }

    // ------------------------------------------------------------------
    // Disassemble — requires the right tool, grants full yield.

    public WorldObjectResult Disassemble(WorldObject obj, Player player)
    {
        if (obj.IsDestroyed)
            return WorldObjectResult.Fail($"The {obj.Name.ToLower()} is already destroyed.");

        var table = DropTableCatalog.For(obj);
        if (table == null || table.DisassembleTool == null)
            return WorldObjectResult.Fail("This can't be taken apart.");

        var tool = FindTool(player, table.DisassembleTool);
        if (tool == null)
            return WorldObjectResult.Fail($"You need a {table.DisassembleTool} to disassemble this.");

        if (obj is Door door && !door.Open)
            return WorldObjectResult.Fail($"You must open the {obj.Name.ToLower()} first.");

        int timeCost = Math.Max(5, (int)Math.Ceiling(obj.MaxHP / 10.0));
        var mats = RollDropTable(table, yieldMod: 1.0);

        obj.HP = 0;
        ApplyDestruction(obj);

        int dropped = DropMaterials(mats, obj.X, obj.Y, obj.Z, player);
        int spilled = SpillContents(obj);

        tool.Durability = Math.Max(0, tool.Durability - 1);

        string spillMsg = spilled > 0 ? $" {spilled} stored item(s) dropped on the ground." : "";
        string gotMsg   = dropped > 0 ? $" ({dropped} materials recovered)"                 : "";

        return new WorldObjectResult(
            success: true,
            message: $"You carefully disassemble the {obj.Name.ToLower()}.{gotMsg}{spillMsg}",
            timeCost: timeCost,
            destroyed: true);
    }

    // ------------------------------------------------------------------
    // Helpers

    /// <summary>Sync the object's tile after destruction so the renderer
    /// shows the new state (open broken door / empty floor).</summary>
    private void ApplyDestruction(WorldObject obj)
    {
        if (obj is Door d)
        {
            ref var tile = ref _world.RefTile(d.X, d.Y, d.Z);
            d.UpdateVisuals(ref tile);
            return;
        }

        if (obj is Furniture f)
        {
            // Clear the tile's object reference and drop back to whatever
            // floor tile the chunk generator painted underneath.
            ref var tile = ref _world.RefTile(f.X, f.Y, f.Z);
            tile.ObjectId = 0;
            // Tile type stays as it was (floor / sidewalk / whatever).
            // Blocking is now gone since ObjectId == 0.
        }
    }

    /// <summary>Spill all stored items from a container onto the ground tile
    /// and clear its pockets. Returns the count dropped. No-op for non-
    /// container objects.</summary>
    private int SpillContents(WorldObject obj)
    {
        if (obj is not Furniture f || !f.IsContainer) return 0;
        int spilled = 0;
        foreach (var pocket in f.Pockets)
        {
            // Copy out before mutating.
            var items = pocket.Contents.ToArray();
            foreach (var it in items)
            {
                pocket.Remove(it);
                _world.AddItemAt(f.X, f.Y, f.Z, it);
                spilled++;
            }
        }
        return spilled;
    }

    private List<(string Name, int Qty, int Qual)> RollDropTable(DropTable? table, double yieldMod)
    {
        var result = new List<(string, int, int)>();
        if (table == null) return result;
        foreach (var m in table.Materials)
        {
            int qty  = _rng.Next(m.MinQ, m.MaxQ + 1);
            int qual = _rng.Next(m.MinQual, m.MaxQual + 1);
            int actualQty = Math.Max(1, (int)(qty * yieldMod));
            result.Add((m.Name, actualQty, (int)(qual * yieldMod)));
        }
        return result;
    }

    /// <summary>
    /// Convert rolled materials into <see cref="Item"/>s, try to stuff them
    /// into the player's inventory, and drop the remainder on the ground.
    /// Returns the number of unique entries that were produced (0 if the
    /// table yielded nothing).
    /// </summary>
    private int DropMaterials(
        List<(string Name, int Qty, int Qual)> mats,
        int x, int y, int z, Player player)
    {
        int produced = 0;
        foreach (var (name, qty, qual) in mats)
        {
            var id = DropNameMap.FamilyIdFor(name);
            Item? item = id != null ? ItemCatalog.Create(id) : null;
            if (item == null)
            {
                // Unknown drop entry — create a bare stackable component so
                // the material still drops instead of being silently lost.
                item = new Item
                {
                    FamilyId = name.ToLowerInvariant().Replace(' ', '_'),
                    Name     = name,
                    Category = "component",
                    Glyph    = '*',
                    Fg       = 214,
                    Weight   = qty * 10,
                    IsComponent = true,
                };
                item.Tags.Add("material");
                item.Tags.Add("salvage");
            }

            item.Quantity = qty;
            if (qual > 0) { item.Durability = Math.Min(item.MaxDurability, qual); }

            // Try player inventory first, fall back to the ground tile.
            var fit = player.Inventory.FindFit(item);
            if (fit != null && fit.TryAdd(item))
                _log($"Recovered {name} x{qty}.");
            else
            {
                _world.AddItemAt(x, y, z, item);
                _log($"{name} x{qty} drops on the ground.");
            }
            produced++;
        }
        return produced;
    }

    /// <summary>
    /// Blunt weapons are good against wood, bad against metal, great against
    /// glass. Sharp weapons are decent across the board. Identical to the
    /// JS table in WorldObjectSystem.getMaterialModifier.
    /// </summary>
    private static double GetMaterialModifier(string material, IReadOnlyList<string> tags)
    {
        bool blunt = tags.Contains("blunt");
        bool sharp = tags.Contains("sharp");

        if (blunt)
        {
            if (material == "wood")  return 1.5;
            if (material == "metal") return 0.5;
            if (material == "glass") return 2.0;
        }
        if (sharp)
        {
            if (material == "wood")  return 1.2;
            if (material == "metal") return 0.7;
            if (material == "glass") return 1.5;
        }
        return 1.0;
    }

    /// <summary>"1d6+2" → average 5. Returns 0 on null/empty.</summary>
    private static int AverageDice(string? dice)
    {
        if (string.IsNullOrEmpty(dice)) return 0;
        int plus = 0;
        string s = dice;
        int pi = s.IndexOfAny(new[] { '+', '-' });
        if (pi > 0)
        {
            plus = int.Parse(s.Substring(pi));
            s = s.Substring(0, pi);
        }
        var parts = s.Split('d');
        if (parts.Length != 2) return plus;
        if (!int.TryParse(parts[0], out int n)) return plus;
        if (!int.TryParse(parts[1], out int f)) return plus;
        // Average of n dN = n * (f+1) / 2, rounded up.
        return (int)Math.Ceiling(n * (f + 1) / 2.0) + plus;
    }

    /// <summary>
    /// Scan equipment and then every pocket reachable from the player for
    /// the first item that matches <paramref name="toolTag"/>. Uses the same
    /// alias groups as JS WorldObjectSystem.isToolType.
    /// </summary>
    private static Item? FindTool(Player player, string toolTag)
    {
        string[] aliases = toolTag switch
        {
            "screwdriver" => new[] { "screwdriver", "multitool" },
            "crowbar"     => new[] { "crowbar", "prybar" },
            "hacksaw"     => new[] { "hacksaw", "saw" },
            "lockpick"    => new[] { "lockpick", "lockpick_set" },
            _             => new[] { toolTag },
        };

        bool Matches(Item it) => aliases.Any(a => it.HasTag(a) || it.FamilyId == a);

        foreach (var it in player.Inventory.EnumerateReachableItems())
            if (Matches(it)) return it;
        return null;
    }
}
